inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -40,376 +1294,440 @@ bool MKLDNNTopKNode::isSupportedOperation(const std::shared_ptr<ngraph::Node>& o
-        SizeVector dst_idx_dims = getChildEdgesAtPort(TOPK_INDEX)[0]->getShape().getStaticDims();
+    for (const auto &df : dataFomats) {
+        addSupportedPrimDesc({{df.first, dataPrecision}, {LayoutType::ncsp, Precision::I32}},
+                             {{df.second, dataPrecision}, {df.second, Precision::I32}},
+                             impl_type);
+    }
+}
 
-        if (dst_idx_dims[axis] != static_cast<size_t>(src_k) || dst_data_dims[axis] != static_cast<size_t>(src_k)) {
-            std::string errorMsg = "Output tensors dimension mismatch";
-            IE_THROW() << errorMsg;
-        }
+void MKLDNNTopKNode::createPrimitive() {
+    auto &dstMemPtr = getChildEdgeAt(TOPK_DATA)->getMemoryPtr();
+    auto &srcMemPtr = getParentEdgeAt(TOPK_DATA)->getMemoryPtr();
+    if (!dstMemPtr || !dstMemPtr->GetPrimitivePtr())
+        IE_THROW() << errorPrefix << " has not allocated destination memory.";
+    if (!srcMemPtr || !srcMemPtr->GetPrimitivePtr())
+        IE_THROW() << errorPrefix << " has not allocate input memory.";
+    if (getSelectedPrimitiveDescriptor() == nullptr)
+        IE_THROW() << errorPrefix << " has nullable preferable primitive descriptor";
+
+    if (getParentEdgeAt(0)->getMemory().GetDesc().hasLayoutType(LayoutType::ncsp)) {
+        layout = TopKLayoutType::topk_ncsp;
+    } else if (getParentEdgeAt(0)->getMemory().GetDesc().hasLayoutType(LayoutType::nspc)) {
+        layout = TopKLayoutType::topk_nspc;
     } else {
-        std::string errorMsg = "Output tensors amount mismatch";
-        IE_THROW() << errorMsg;
+        layout = TopKLayoutType::topk_blocked;
     }
 
-    if (src_dims[axis] < static_cast<size_t>(src_k))
-        src_k = src_dims[axis];
+    auto selectedPD = getSelectedPrimitiveDescriptor();
+    auto data_type = MKLDNNExtensionUtils::IEPrecisionToDataType(selectedPD->getConfig().inConfs[TOPK_DATA].desc->getPrecision());
+    data_size = MKLDNNExtensionUtils::sizeOfDataType(data_type);
 
-    SizeVector in_dims = getParentEdgeAt(TOPK_DATA)->getShape().getStaticDims();
+    topk_innermost = false;
+    if ((layout == TopKLayoutType::topk_ncsp && axis == dst_dims_size - 1) ||
+       ((layout == TopKLayoutType::topk_nspc || layout == TopKLayoutType::topk_blocked) && axis == 1))
+        topk_innermost = true;
 
-    if (src_k == 1) {
-        if (is_last_dim) {
-            if (mode_max)
-                top1<std::greater>(src, dst_data, dst_idx, in_dims);
-            else
-                top1<std::less>(src, dst_data, dst_idx, in_dims);
-        } else {
-            if (mode_max)
-                top1_axis<cmpgt_ps, std::greater>(src, dst_data, dst_idx, in_dims);
-            else
-                top1_axis<cmplt_ps, std::less>(src, dst_data, dst_idx, in_dims);
-        }
+    if (mayiuse(cpu::x64::avx512_common)) {
+        blk_size = 16;
+        count_xmm = 16; // only 16 vector registers are valid in sse instructions
```

Comment:
Is it? Seems like AVX512VL allows to work with all 32 registers as XMM ones.

Reply:
I think your statement is true. Yet the main concern is that there are some commonly used jit functions(eg. uni_vmovdqu and uni_vcvtdq2ps) in topk node as well as in emmiters. As far as I'm concerned, they call sse2 instrucions(eg. movdqu and cvtdq2ps respectively) if the input is xmm. Since the instructions are sse2 based, they are not allowed to refer to register 16~31 even the isa is avx512. In fact, when I change the value to 32, I get unexpected test results. Further discussion is always welcomed, in case I have any misunderstanding here. Thanks Dmitry!